docs/proposals/network-policy.md

```diff
@@ -0,0 +1,259 @@
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NetworkPolicy
+
+## Abstract
+
+A proposal for implementing a new resource - NetworkPolicy - which
+will enable definition of ingress policies for selections of pods.
+
+The design for this proposal has been created by, and discussed
+extensively within the Kubernetes networking SIG.  It has been implemented
+and tested using Kubernetes API extensions by various networking solutions already.
+
+In this design, users can create various NetworkPolicy objects which select groups of pods and
+define how those pods should be allowed to communicate with each other.  The
+implementation of that policy at the network layer is left up to the
+chosen networking solution.
+
+> Note that this proposal does not yet include egress / cidr-based policy, which is still actively undergoing discussion in the SIG.  In general,
+egress is expected to augment this proposal rather than modify it, with one exception (noted below).
+
+## Implementation
+
+The implmentation in Kubernetes consists of:
+- A v1beta NetworkPolicy API object
+- A field on the `Namespace` object to control isolation, to be developed as an annotation for now, called `net.beta.kubernetes.io/network-isolation`.
+
+### Namespace changes
+
+> Note that this section is subject to some change with the addition of egress policy as mentioned above.
+
+The `Namespace` object can be augmented with the annotation `net.beta.kubernetes.io/network-isolation` which will take one of the following values:
+
+When `net.beta.kubernetes.io/network-isolation=true` is set on a namespace:
+- Pods in that Namespace will not be accessible from any other source, even sources within that namespace, unless explicitly allowed by a NetworkPolicy object.
+- Pods in that Namespace will be able to access any other destination.
+
+When `net.beta.kubernetes.io/network-isolation=false` is set on a namespace:
+- Pods in that Namespace will be accessible from any other source.
+- Pods in that Namespace will be able to access any other destination.
+
+This field will default to `false` if not specified.
+
+### NetworkPolicy Go Definition 
+
+Once the namespace is isolated, a mechanism to selectively allow traffic into the namespace and between pods within
+the namespace is required.  That is accomplished through ingress rules on `NetworkPolicy`
+objects (of which there can be multiple in a single namespace).  Pods selected by
+one or more NetworkPolicy objects should allow any incoming connections that match any
+ingress rule on those NetworkPolicy objects, per the network plugin’s capabilities.
```

Comment:
Do we allow a network plugin to not be capable of fully implementing this spec?  I can think of a plausible motivation for that: if the operator chooses to use userspace proxies for k8s Services, that make full implementation of this spec extra difficult --- but it poses no extra problem for NetworkPolicy objects that do not discriminate on client.  Do we allow an operator to use a network plugin that does not implement this spec at all?

If we allow less than full implementation of this spec, what happens to input that can not be fully implemented?  Is it rejected at validation?  Is there a way to merely issue a warning at input time?  Or should a warning appear in a status block?

I assume we all came at this from a security and restriction point of view.  But if we contemplate less than full implementation then we need a more nuanced story.  I could imagine this point of view: a client creates NetworkPolicy objects to describe the connectivity it needs; the cluster's job is to provide at least that much connectivity; a good cluster will reject some unneeded traffic; a really good cluster will reject all unneeded traffic; the more concerned the user is about security, the better the cluster he will choose to run his application.